ouroboros-consensus/src/ouroboros-consensus/Ouroboros/Consensus/Genesis/Governor.hs

```diff
@@ -93,4 +114,147 @@ updateLoEFragStall k getCandidates =
+      pure peer0
+
+    loeIntersectionSlot = AF.headSlot loeFrag
+    -- exclusive last slot in the Genesis window
+    endOfGenesisWindow =
+        succWithOrigin loeIntersectionSlot + SlotNo sgen
+
+    dropBeyondGenesisWindow =
+      AF.takeWhileOldest ((< endOfGenesisWindow) . blockSlot)
+
+    competingFrags =
+      Map.map dropBeyondGenesisWindow candidateSuffixes
+
+    -- Length of an interval with inclusive lower bound @a@ and exclusive upper
+    -- bound @b@.
+    intervalLength a b
+      | a <= b    = b - a
+      | otherwise = 0
+
+    showBounds :: (AnchoredFragment (Header blk), Bool, Word64, Word64) -> String
+    showBounds (_, more, lower, upper) = show lower ++ "/" ++ show upper ++ "[" ++ (if more then "+" else " ") ++ "]"
+
+{- TODO: convert this scribble into a useful explanatory diagram, illustrating the
+        density calculation below
+
+            |--------|
+
+    frag1: A - B - C - D - ...            <- exact
+            \
+    frag2:     E           claimed tip: E <- exact
+
+-}
+
+-- | Run the Genesis disconnection logic once.
+--
+--   * Maintain and update the LoE.
+--
+--   * Disconnect from peers with inferior density.
+updateLoEFragGenesis ::
+     forall m blk peer.
+     ( IOLike m
+     , Ord peer
+     , Show peer
+     , LedgerSupportsProtocol blk
+     )
+  => TopLevelConfig blk
+  -> Tracer m String
+  -> STM m (Map peer (AnchoredFragment (Header blk)))
+  -> STM m (Map peer (ChainSyncClientHandle m blk))
+  -> UpdateLoEFrag m blk
+updateLoEFragGenesis cfg tracer getCandidates getHandles =
+  UpdateLoEFrag $ \ curChain immutableLedgerSt setLoEFrag -> do
+    (candidateSuffixes, handles, theirTips, loeFrag) <- atomically $ do
+      candidates <- getCandidates
+      handles <- getHandles
+      let
+        (loeFrag, candidateSuffixes) =
+          sharedCandidatePrefix (configSecurityParam cfg) curChain candidates
+      setLoEFrag loeFrag
+      -- Obtaining these before calling setLoEFrag appeared to result in a wrong state,
+      -- but maybe that was a fluke.
+      theirTips <-
+        fmap (Map.mapMaybe id) $ flip Map.traverseWithKey candidateSuffixes $ \peer _ ->
+          cschTheirTip (handles Map.! peer)
+      pure (candidateSuffixes, handles, theirTips, loeFrag)
+
+    let
+      -- TODO: use a forecasted ledger view for the intersection
+      -- slot (tip of LoE frag).
+      sgen = computeGenesisWindow (configLedger cfg) (ledgerState immutableLedgerSt)
+      (losingPeers, boundsDesc) =
+        densityDisconnect sgen (configSecurityParam cfg) candidateSuffixes theirTips loeFrag
+
+    trace ("Density bounds: " ++ boundsDesc)
+    trace ("New candidate tips: " ++ showPeers (showTip <$> Map.map AF.headPoint candidateSuffixes))
+    trace ("Losing peers: " ++ show losingPeers)
+    trace ("Setting loeFrag: " ++ show (AF.headAnchor loeFrag))
+
+    for_ losingPeers $ \peer -> do
+      trace ("Killing peer: " ++ show peer)
+      cschKill (handles Map.! peer)
```

Comment:
This kills the ChainSync client and allows the peer to continue sending blocks via BlockFetch. Is not killing `BlockFetch` an omission?

Reply:
The diffusion layer should entirely disconnect from a peer when an exception is thrown in any miniprotocol for that peer, so it should be enough to kill one miniprotocol. Are you observing something to the contrary?

Reply:
Yes. In the test "serve adversarial branches" the disconnected adversary still sends blocks.

Reply:
Ah, I guess we don't invoke the actual network diffusion layer, so this is probably something that should be fixed in the Genesis peer simulator infrastructure.

Reply:
This was addressed in #981.